Prime players to enable autoplay when out of focus

CookPete opened the pull request:
My version of https://github.com/CookPete/react-player/pull/12 that attempts to fix [autoplay issues](https://github.com/CookPete/react-player/issues/7) when `ReactPlayer` is not in a focused tab. In short:

* Render all players at once, but only give `props` to the player that can play the current `url`
* Each player then sets `display: none` on itself if `url` is not present
* When `ReactPlayer` mounts, if `YouTube` or `Vimeo` players don't have a `url`, then `prime` the player by playing a short, silent video whilst hidden. This allows the players to autoplay any URLs given later on, even if not in a focused tab.

@Fauntleroy, any thoughts?

Fauntleroy commented:
I was thinking of making the priming functionality manual by using a configuration parameter:

```js
const reactPlayerProps = {
  youtubeConfig: {
    preload: true
  },
  vimeoConfig: {
    preload: true
  }
};
<ReactPlayer {...reactPlayerProps} />
```

This way we could avoid loading those APIs for those who don't intend on using them.

CookPete commented:
Updated. Moved priming logic from `Base` into `Youtube` and `Vimeo` players, so they can check their config and decide whether to prime or not. Should probably align the config name (currently `preload`) with the variable name (currently `priming`) and add some documentation. It also makes prop logic in `ReactPlayer` [rather awkward](https://github.com/CookPete/react-player/blob/fix-autoplay-prime/src/ReactPlayer.js#L52-L60) so there's potential for improvement there.

CookPete commented:
Updated. Now using slightly less ambiguous `preloading` var name

CookPete commented:
Updated (again) with a short explanation in `README.md`

CookPete commented:
Looking to merge this soon even if it's a bit weird.. @Fauntleroy are you happy enough?

Fauntleroy commented:
@CookPete I'll take a look at this soon to ensure that it works properly. My initial test earlier this week was having some problems. Can you check and see if `youtubeConfig` makes it all the way to the player?

CookPete commented:
> Can you check and see if youtubeConfig makes it all the way to the player?
 
Ah yeah that should have been sorted by https://github.com/CookPete/react-player/commit/52574f8aafffded6392016b8ede7b6c9d9d75fc1

Fauntleroy commented:
Something is definitely wrong when I try to use this in my application. I'm going to try a few things and see if I can resolve the problem.

Fauntleroy commented:
Autoplay seems to behave differently than it did before, but with `playing` and `youtubeConfig` i've managed to get that settled.

Volume, however, **is** broken. Looks like you just need to make sure volume is included in `shouldComponentUpdate`:

```js
shouldComponentUpdate (nextProps) {
    return (
      this.props.url !== nextProps.url ||
      this.props.playing !== nextProps.playing ||
      this.props.volume !== nextProps.volume
    )
  }
```

Fauntleroy commented:
I'm also running into `Uncaught TypeError: Cannot read property 'getFractionLoaded' of undefined` when using the SoundCloud player, though I'm trying to figure out how it's even happening.

Fauntleroy commented:
The issue with `getFractionLoaded` can be solved by checking to see if `this.refs.player` exists during the if statement wrapping the contents of `progress` in `ReactPlayer`:

```js
progress = () => {
    if (this.props.url && this.refs.player) {
      let progress = {}
      const loaded = this.refs.player.getFractionLoaded()
      const played = this.refs.player.getFractionPlayed()
```

Fauntleroy commented:
I'm also having issues rendering all of the players inside my application. For some reason only YouTube will bother to render. Here's what my React inspector is showing me:

<img width="452" alt="screen shot 2015-12-18 at 11 55 52 pm" src="https://cloud.githubusercontent.com/assets/507047/11912099/09133796-a5e3-11e5-89fb-cb42703bf245.png">

I imagine something horrible is happening to important parts of the code when minification occurs. Maybe all of the component names are being destroyed, and that's causing a few things to break here and there?

CookPete commented:
Wow, thanks very much @Fauntleroy for the testing! I've added [some more fixups](https://github.com/CookPete/react-player/compare/2dc18f7ef01341d84a7ff5951605106448643958...11b589d279b74776c0aa539cd832297e5d25636d) to hopefully solve most of the issues.

* Separate `load` and `play` methods are now used to clarify between actually *loading* media and *playing* it. Media will `play` after `load` if `props.playing` is true at the time. This should clear up the autoplay issues we've been seeing, and solidify the logic between loading and playing. For example, you can now keep `playing: false` but click through the demo and load media ready for playing whenever `playing` becomes `true.
  * This actually makes `autoplay` config overrides sort of obsolete/pointless. If you have `playing: true`, media should autoplay no matter what, and if `playing` is `false` then it shouldn't. Any other `autoplay` settings would just confuse the logic and the players' state.
* I've added the `volume` and `getFractionLoaded` fixes you suggested. Good spot!
* I need to look more at the minified problems. What exact compressor/options are you using?

Fauntleroy commented:
@CookPete it might not have to do with minification. I seem to remember being able to replicate the issue without minified code, but I can't remember exactly how that came about. I'll look into it more soon.

CookPete commented:
I'm gonna merge this in an attempt to avoid conflicts with other fixes. Hopefully all problems (other than the weird minified issue) have been sorted
